src/gc/boehm.cr

```diff
@@ -132,13 +132,22 @@ module GC
     # By default the GC warns on big allocations/reallocations. This
     # is of limited use and pollutes program output with warnings.
     LibGC.set_warn_proc ->(msg, v) do
-      format_string = String.new(msg)
-      unless format_string.starts_with?("GC Warning: Repeated allocation of very large block")
-        LibC.printf format_string, v
+      unless GC.starts_with?(msg, "GC Warning: Repeated allocation of very large block")
+        LibC.printf msg, v
       end
     end
   end
 
+  # :nodoc:
+  # This method implements `String#starts_with?` without allocating a `String` (#11728)
+  def self.starts_with?(string_pointer, start)
+    start.to_slice.each_with_index do |chr, i|
+      return false unless chr == string_pointer[i]
```

Comment:
Shouldn't this watch for `string_pointer[i] == 0` and then break the loop?

Reply:
I think once string_pointer[I] is zero it won't match the char in the expected string, so the loop will break.

Reply:
True. Still very implicit way of dealing with this that relies on non-attacker controlled input in `start`, so might be worth an explicit check or an explanatory comment.

Reply:
This method is customized for this specific use case and we're controlling the value of `start` (which governs the iteration length). So this should be completely safe.

Reply:
Until somebody starts to copy or reuse it without much thinking. As said I'm fine with a comment but I would not leave this as an uncommented example in a codebase without doubt many use as an inspiration :)

Reply:
>  we're controlling the value of `start` (which governs the iteration length)

I don't think this matters. It governs the iteration length, but what happens if the string is longer than `start`? Then almost by chance the iteration stops because C strings are guaranteed to end with a final null byte. But, as @jhass says, it's not immediately clear.

I would much rather prefer to use `LibC.strlen` and using `LibC.strncmp`.

Reply:
Okay, I'll refactor to use `strlen`.

We currently don't have bindings for `strncmp`, but I believe we don't need that. As soon as we know the length of the string, we can use a slice for the comparison.